### DISCUSSION TITLE
How to create a log output widget with auto scroll to bottom?

### DISCUSSION THREAD

**Discussion by AllenDang** [slint-ui/slint#2301] [Q&A] [answered]
I need a widget to display runtime log which support auto scroll to bottom when new log string is added.

Ideally, TextEdit or ScrollView+Text might have some function like "Scroll To XY", I could invoke it to control the scroll position programable, but I didn't find any clue from document and examples.

Any hint?

**Comment by ogoffart** [comment, marked as answer]
That's a feature that make sense. I wonder what API it should have.

Right now, you will have to set the content-y manually to go to the end like so:
```
                log_view.viewport-y = min(0px, - log_view.viewport-height + log_view.visible-height);
```

[Full example](https://slint-ui.com/releases/0.3.5/editor/?snippet=import+%7B+Button%2C+VerticalBox+%2C+TextEdit%2C+LineEdit%7D+from+%22std-widgets.slint%22%3B%0Aexport+component+Demo+%7B%0A++++height%3A+200px%3B%0A++++width%3A+300px%3B%0A++++property+%3Cint%3E+counter%3B%0A++++VerticalBox+%7B%0A++++++++log_view+%3A%3D+TextEdit+%7B%0A++++++++++++text%3A+%22Log+here%5Cnline1%5Cnlinz2%5Cnline3%5Cnline4%5Cnline5%5Cnline6%5Cnblahblah%5Cnxxx%22%3B+%0A++++++++%7D%0A++++++++le+%3A%3D+LineEdit+%7B+text%3A+%22log+to+add%22%3B+%7D%0A++++++++Button+%7B++++%0A++++++++++++text%3A+%22Add%22%3B%0A++++++++++++clicked+%3D%3E+%7B+%0A++++++++++++++++counter+%2B%3D+1%3B%0A++++++++++++++++log_view.text+%2B%3D+%22%5Cn%22+%2B+le.text%3B%0A++++++++++++++++le.text+%3D+%22to+add+%22+%2B+counter%3B%0A%0A++++++++++++++++%2F%2F+GO+TO+THE+END%0A++++++++++++++++log_view.viewport-y+%3D+min%280px%2C+-+log_view.viewport-height+%2B+log_view.visible-height%29%3B%0A++++++++++++%7D+%0A++++++++%7D%0A++++%7D%0A%7D%0A)

**Reply by acgxv** [reply]
+1.

An issue to track this would be better.

**Comment by nic-obert** [comment]
Are there any plans for an out-of-the-box autoscroll feature? I currently have to manually implement a scroll-to-bottom function like @ogoffart  suggested and manually invoke it from the backend code every time I want to add a new output line to a `ListView`.
It would be nice to leave UI functionality, like automatic scroll-to-bottom, out of the backend code.

**Reply by igiona** [reply]
You can kind-of do this fully on the UI side:

```
                    autoscroll_timer := Timer {
                        interval: 100ms;
                        running: autoscroll.checked;
                        triggered() => {
                            log_view.viewport-y = min(0px, - log_view.viewport-height + log_view.visible-height);
                        }
                    }

                    autoscroll := CheckBox {
                        text: "Autoscroll";
                        checked: true;
                    }
```
Of course it's a bit a waste of resources since the timer runs even if the list never changes .... 
An out-of-the box solution would be much appreciated ;)

**Comment by werner291** [comment]
Hi! I was looking at this pattern for a chat app, starting from the timer-based polling shape upthread, and ran into the harder version of the problem: how to keep auto-scroll-to-bottom when the user is allowed to scroll up at any time. The trick I found was to stop trying to tell "user scroll" apart from "layout shift". The observation: **the only direction layout-induced motion can move `viewport-y` is downward** (more negative, in Slint's `vp-y ≤ 0` convention) when content is appended. So any upward motion is the user, by definition.

That collapses the whole thing to a three-rule state machine on a polling Timer:

1. If `viewport-y` increased since the last poll → `sticky = false`.
2. If `viewport-y` is at or below the bottom + a small margin → `sticky = true`.
3. If `sticky` and `viewport-height` grew since the last poll → write `viewport-y = height - viewport-height`.

Packaged as a reusable component:

```slint
export component StickyBottomList inherits Rectangle {
    in property <length> up_tolerance: 0px;
    in property <length> bottom_margin: 16px;
    in property <duration> poll_interval: 100ms;
    out property <bool> sticky_to_bottom: true;
    out property <bool> has_unseen_growth: false;
    public function snap_to_bottom() { /* set sticky, write vp-y, clear */ }

    private property <length> prev_vp_y: 0;
    private property <length> prev_vp_h: 0;

    listview := Flickable {
        VerticalLayout { @children }
    }

    Timer {
        interval: root.poll_interval;
        running: true;
        triggered => {
            if (listview.viewport-y > prev_vp_y + root.up_tolerance) {
                sticky_to_bottom = false;
            }
            if (listview.viewport-y
                <= (listview.height - listview.viewport-height) + root.bottom_margin) {
                sticky_to_bottom = true;
                has_unseen_growth = false;
            }
            if (listview.viewport-height > prev_vp_h) {
                if (sticky_to_bottom) {
                    listview.viewport-y = listview.height - listview.viewport-height;
                } else {
                    has_unseen_growth = true;
                }
            }
            prev_vp_y = listview.viewport-y;
            prev_vp_h = listview.viewport-height;
        }
    }
}
```

The `has_unseen_growth` output is handy for driving a "new messages below" pill that calls `snap_to_bottom()`.

**Comment by RyanHope** [comment]
any updates on this?
